pennylane/pauli/pauli_arithmetic.py

```diff
@@ -0,0 +1,279 @@
+    """Immutable dictionary used to represent a Pauli Word.
+    Can be constructed from a standard dictionary.
+
+    >>> w = PauliWord({"a": X, 2: Y, 3: Z})
+    """
+
+    def __missing__(self, key):
+        """If the wire is not in the Pauli word,
+        then no operator acts on it, so return the Identity."""
+        return I
+
+    def __init__(self, mapping):
+        """Strip identities from PauliWord on init!"""
+        mapping = {wire: op for wire, op in mapping.items() if op != I}
+        super().__init__(mapping)
+
+    def __copy__(self):
+        """Copy the PauliWord instance."""
+        return PauliWord(dict(self.items()))
+
+    def __setitem__(self, key, item):
+        """Restrict setting items after instantiation."""
+        raise NotImplementedError
+
+    def update(self, __m, **kwargs) -> None:
+        """Restrict updating PW after instantiation."""
+        raise NotImplementedError
+
+    def __hash__(self):
+        return hash(frozenset(self.items()))
+
+    def __mul__(self, other):
+        """Multiply two Pauli words together using the matrix product if wires overlap
+        and the tensor product otherwise.
+
+        Args:
+            other (PauliWord): The Pauli word to multiply with
+
+        Returns:
+            result(PauliWord): The resulting operator of the multiplication
+            coeff(complex): The complex phase factor
+        """
+        c_self, c_other = (copy(self), copy(other))
+        result, iterator, swapped = (
+            (dict(c_self), c_other, False)
+            if len(self) > len(other)
+            else (dict(c_other), c_self, True)
+        )
+        coeff = 1
+
+        for wire, term in iterator.items():
+            if wire in result:
+                factor, new_op = (
+                    mul_map[term][result[wire]] if swapped else mul_map[result[wire]][term]
+                )
+                if new_op == I:
+                    del result[wire]
+                else:
+                    coeff *= factor
+                    result[wire] = new_op
+            elif term != I:
+                result[wire] = term
+
+        return PauliWord(result), coeff
+
+    @property
+    def wires(self):
+        """Track wires in a PauliWord."""
+        return set(self)
+
+    def to_mat(self, wire_order, format="dense"):
+        """Given a Pauli word, get the matrix representation.
+
+        KeywordArgs:
+            wire_order (iterable or None): The order of qubits in the tensor product.
+            format (str): The format of the matrix ("dense" by default), if not a dense
+                matrix, then the format for the sparse representation of the matrix.
+
+        Returns:
+            (Union[NumpyArray, ScipySparseArray]): Matrix representation of the Pauliword
+
+        Raises:
+            ValueError: Can't get the matrix of an empty PauliWord.
+        """
+        if len(self) == 0:
+            raise ValueError("Can't get the matrix of an empty PauliWord.")
```

Comment:
Identity?

Reply:
An empty PauliSentence and PauliWord act kind of like the identity on all wires. So I would be hesitant to return a fixed dimension matrix here since the behaviours wouldn't match. I think having a error is fine 👍🏼

Reply:
Maybe `if len(self) == 0 and wire_order is None`?
If we have a `wire_order`, then we return an identity of the right size as wire order.